pkg/onit/app.go

```diff
@@ -0,0 +1,285 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package onit
+
+import (
+	"fmt"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"time"
+)
+
+// GetSApps returns a list of apps deployed in the cluster
+func (c *ClusterController) GetApps() ([]string, error) {
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"app": "onos", "type": "app"}}
+	appList, err := c.kubeclient.AppsV1().Deployments(c.clusterID).List(metav1.ListOptions{
+		LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		return nil, err
+	}
+
+	apps := make([]string, len(appList.Items))
+	for i, app := range appList.Items {
+		apps[i] = app.Name
+	}
+	return apps, nil
+}
+
+// setupApp creates an app
+func (c *ClusterController) setupApp(name string, config *AppConfig) error {
+	if err := c.createAppConfigMap(name, config); err != nil {
+		return err
+	}
+	if err := c.createAppService(name); err != nil {
+		return err
+	}
+	if err := c.createOnosAppDeployment(name, config.Image); err != nil {
+		return err
+	}
+	if err := c.awaitOnosAppDeploymentReady(name); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createAppConfigMap creates an app configuration
+func (c *ClusterController) createAppConfigMap(name string, config *AppConfig) error {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: c.clusterID,
+		},
+		Data: map[string]string{},
+	}
+	_, err := c.kubeclient.CoreV1().ConfigMaps(c.clusterID).Create(cm)
+	return err
+}
+
+// createOnosAppDeployment creates an app Deployment
+func (c *ClusterController) createOnosAppDeployment(name string, image string) error {
+	nodes := int32(1)
+	zero := int64(0)
+	dep := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: c.clusterID,
+			Labels: map[string]string{
+				"app":  "onos",
+				"type": "app",
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &nodes,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app":  "onos",
+					"type": "app",
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":      "onos",
+						"type":     "app",
+						"resource": name,
+					},
+					Annotations: map[string]string{
+						"seccomp.security.alpha.kubernetes.io/pod": "unconfined",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            name,
+							Image:           image,
```

Comment:
I would suggest to use c.ImageName function like in other subsystems to pass a name and a tag. In the command itself we need to pass a name and an optional flag for the tag (look at onit create cluster command as an example) of the image in case we need to deploy a specific version of the app.